File: ocaml/examples/hello.ml
(** hello.ml — minimal example of the ktdeque OCaml library.

    Shows the idiomatic public interface: [Ktdeque.Deque] (the §4
    real-time deque) and [Ktdeque.Cadeque] (adds O(1) concatenation).
    Every operation is purely functional and worst-case O(1); values are
    persistent, so a deque can be "forked" and both branches used.

    Build (after `opam install .` from the repo root):

        ocamlfind ocamlopt -package ktdeque -linkpkg hello.ml -o hello

    Or in-tree: `dune exec ocaml/examples/hello.exe`. *)

open Ktdeque

let show label xs =
  Printf.printf "%s = [%s]\n" label
    (String.concat "; " (List.map string_of_int xs))

let () =
  (* ---- Deque: push at front, inject at back, pop/eject the ends ---- *)
  let d = Deque.of_list [10; 20] in     (* front = 10, back = 20 *)
  let d = Deque.push 30 d in            (* [30; 10; 20] *)
  show "d" (Deque.to_list d);

  (* Persistence: a derived deque leaves the original intact. *)
  let branch = Deque.inject d 40 in     (* [30; 10; 20; 40] *)
  show "branch" (Deque.to_list branch);
  show "d (still)" (Deque.to_list d);

  (match Deque.pop d with
   | Some (x, _) -> Printf.printf "front of d: %d\n" x
   | None -> ());
  (match Deque.eject d with
   | Some (_, x) -> Printf.printf "back of d:  %d\n" x
   | None -> ());

  (* ---- Cadeque: the same, plus worst-case O(1) concat ---- *)
  let a = Cadeque.of_list [1; 2; 3] in
  let b = Cadeque.of_list [4; 5; 6] in
  let c = Cadeque.concat a b in         (* O(1), regardless of size *)
  show "concat a b" (Cadeque.to_list c)

File: c/examples/hello.c
/* hello.c — minimal example of using the ktdeque C library.
 *
 * Walks through the four core operations (push / inject / pop / eject)
 * plus a brief persistence demonstration.  Every line is annotated.
 *
 * For the algorithm intuition (why this deque is worst-case O(1) per
 * op when the natural "spill on overflow" approach would be O(log n)),
 * read kb/spec/why-bounded-cascade.md.  For the public C surface,
 * see ../include/ktdeque.h.
 *
 * Build (after `make install` from c/):
 *
 *     cc hello.c -lktdeque -o hello
 *
 * Or via pkg-config:
 *
 *     cc hello.c $(pkg-config --cflags --libs ktdeque) -o hello
 *
 * Or directly against the in-tree build:
 *
 *     cc -I../include -L.. hello.c -lktdeque -o hello
 *
 * Then: ./hello
 */
#include <ktdeque.h>
#include <stdio.h>

static void print_long(kt_elem e, void* ctx) {
    int* first = (int*)ctx;
    if (!*first) printf(", ");
    *first = 0;
    printf("%ld", *(long*)e);
}

int main(void) {
    /* Storage backing for the values we push.  ktdeque does not copy or
     * own values — it stores raw pointers, so the storage must outlive
     * the deque. */
    long values[] = { 10, 20, 30, 40 };

    kt_deque d = kt_empty();
    d = kt_push  (kt_base(&values[0]), d);   /* front -> 10 */
    d = kt_inject(d, kt_base(&values[1]));   /* back  <- 20 */
    d = kt_push  (kt_base(&values[2]), d);   /* front -> 30 */
    /* Deque is now [30, 10, 20]. */

    /* Persistence: branching d gives independent derivatives. */
    kt_deque branch = kt_inject(d, kt_base(&values[3]));  /* [30, 10, 20, 40] */

    printf("d      = ["); { int f = 1; kt_walk(d,      print_long, &f); } printf("]\n");
    printf("branch = ["); { int f = 1; kt_walk(branch, print_long, &f); } printf("]\n");
    printf("|d| = %zu, |branch| = %zu\n",
           kt_length(d), kt_length(branch));

    /* Drain d via pop+eject. */
    kt_elem out; int nonempty;
    d = kt_pop  (d, &out, &nonempty); printf("pop:   %ld\n", *(long*)out);
    d = kt_eject(d, &out, &nonempty); printf("eject: %ld\n", *(long*)out);
    d = kt_pop  (d, &out, &nonempty); printf("pop:   %ld\n", *(long*)out);
    d = kt_pop  (d, &out, &nonempty); printf("empty? %d\n", !nonempty);
    return 0;
}
